crates/sui-indexer-alt-jsonrpc-proxy/src/handlers.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::cursor::{transform_json_body, update_pagination_cursor_state, PaginationCursorState};
+use crate::metrics::AppMetrics;
+use axum::http::HeaderValue;
+use axum::{
+    body::Body,
+    extract::{Request, State},
+    http::request::Parts,
+    http::StatusCode,
+    response::Response,
+};
+use bytes::Bytes;
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::Instant;
+use tracing::{debug, info, warn};
+use url::Url;
+
+#[derive(Clone)]
+pub struct AppState {
+    client: reqwest::Client,
+    fullnode_address: Url,
+    unsupported_methods: HashSet<String>,
+    cursor_state: Arc<PaginationCursorState>,
+    metrics: AppMetrics,
+}
+
+impl AppState {
+    pub fn new(
+        client: reqwest::Client,
+        fullnode_address: Url,
+        unsupported_methods: HashSet<String>,
+        cursor_state: Arc<PaginationCursorState>,
+        metrics: AppMetrics,
+    ) -> Self {
+        Self {
+            client,
+            fullnode_address,
+            unsupported_methods,
+            cursor_state,
+            metrics,
+        }
+    }
+}
+
+pub async fn proxy_handler(
+    State(state): State<AppState>,
+    request: Request<Body>,
+) -> Result<Response, (StatusCode, String)> {
+    let (mut parts, body) = request.into_parts();
+
+    let mut body_bytes = match axum::body::to_bytes(body, 10 * 1024 * 1024).await {
+        Ok(bytes) => bytes,
+        Err(e) => {
+            warn!("Failed to read request body: {}", e);
+            return Ok(Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from("Failed to read request body"))
+                .unwrap());
+        }
+    };
+
+    match parts
+        .headers
+        .get("Client-Request-Method")
+        .and_then(|h| h.to_str().ok())
+    {
+        Some(method) if state.unsupported_methods.contains(method) => {
+            debug!("Dropping {method} request");
+            Ok(Response::builder()
+                .status(StatusCode::OK)
+                .body(Body::from(format!("Dropped {method} request")))
+                .unwrap())
+        }
+        _ => {
+            let mut json_body = match serde_json::from_slice::<serde_json::Value>(&body_bytes) {
+                Ok(json_body) => json_body,
+                Err(_) => {
+                    debug!("Failed to parse request body as JSON");
+                    return Ok(proxy_request(&state, parts, body_bytes).await?.0);
+                }
+            };
```

Comment:
```suggestion
            let Ok(mut json_body) = serde_json::from_slice::<serde_json::Value>(&body_bytes) else {
                debug!("Failed to parse request body as JSON");
                return Ok(proxy_request(&state, parts, body_bytes).await?.0);
            };
```

Is it correct to proxy the request if we can't parse it as JSON? I thought we needed a positive confirmation that we know what the request is.